WcaOnRails/app/controllers/delegate_reports_controller.rb

```diff
@@ -0,0 +1,56 @@
+class DelegateReportsController < ApplicationController
+  before_action :authenticate_user!
+
+  before_action -> { redirect_unless_user(:can_view_delegate_report?, competition_from_params) }, only: [
+    :show,
+  ]
+  before_action -> { redirect_unless_user(:can_edit_competition_report?, competition_from_params) }, only: [
+    :edit,
+    :update,
+  ]
+
+  private def competition_from_params
+    if params[:competition_id]
+      competition = Competition.find(params[:competition_id])
+    else
+      report = DelegateReport.find(params[:id])
+      competition = report.competition
+    end
+    competition
```

Comment:
I think it would be a bit weird, why wouldn't we let the if statement exist here =)
```ruby
if params[:competition_id]
  Competition.find(params[:competition_id])
else
  DelegateReport.find(params[:id]).competition
end
```